# Sanity check

#!/bin/bash
#****************************************
# To test the repository, use phoenix:

# Login to the cluster:

# ssh ntavakoli6@login-phoenix.pace.gatech.edu 
# ssh ntavakoli6@login-phoenix-3.pace.gatech.edu

# pace-quota
# To submit an interactive job:

#  qsub -I -q inferno -A GT-saluru8-CODA20 -l nodes=1:ppn=24,mem=100gb,walltime=96:00:00
#*****************************************

# # Change this line according to your projcet directory
# cd /storage/coda1/p-saluru8/0/ntavakoli6/hged
# #****  NOTE:  This part needs to be done only once to download vcf and the required reference
# chmod +x download_vcf_and_ref.sh
# ./download vcf_and_ref.sh
# chmod +x download_sw_dependencies.sh
# ./download_sw_dependencies.sh
# ***************************************************************************************

# Change this line according to your projcet directory
cd /storage/coda1/p-saluru8/0/ntavakoli6/hged
#----------------------------------------------------------------

project_dir=$(pwd)  #project top-level directory

cd ${project_dir}/data
DATA=$(pwd)

cd ${project_dir}/software
software_dir=$(pwd)

cd ${project_dir}
bcftools=${software_dir}/bcftools-1.9/bcftools
vcftools=${software_dir}/vcftools-0.1.16/bin/vcftools
tabix=${software_dir}/htslib-1.12/tabix
bgzip=${software_dir}/htslib-1.12/bgzip
samtools=${software_dir}/samtools-1.12/samtools

cd ${DATA} 

id=22
sample=HG00096
ref=hs37d5
MyVariants=${DATA}/chr${id}.vcf.gz

cat chr22_sample.txt | head -10
# HG00096
# HG00097
# HG00099
# HG00100
# HG00101
# HG00102
# HG00103
# HG00105
# HG00106
# HG00107

# Get variant positions for chr22
cat chr22_indels.frq.count | grep 51214426
# 22	51214426	2	5008	CGT:4994	C:14

$bcftools view -H  $MyVariants  | grep 51214426


# get the substring of length alpha from the second haplotype of the sample HG00096
$samtools faidx hs37d5.fa 22:51214426-51214430 | $bcftools consensus -s HG00096 -H 2 chr${id}_sample_22.vcf.gz.gz >  chr${id}_sample_${sample}.fa


id=22
sample=HG00096
ref=hs37d5
# nl is used to get line number
 $bcftools view -H chr_22_HG00096.vcf.gz | nl | grep "1|0"
 #380390 22		29104940 CAA	C 1|0

$samtools faidx hs37d5.fa 22:29104940-29104950
#CAAAAAAAAAA

$samtools faidx hs37d5.fa 22:29104940-29104950| $bcftools consensus -s ${sample} -H 1 chr_${id}_${sample}.vcf.gz 
#CAAAAAAAA


$samtools faidx hs37d5.fa 22:29104940-29104950| $bcftools consensus -s ${sample} -H 2 chr_${id}_${sample}.vcf.gz 
#CAAAAAAAAAA

# We have two variant positions
16050678 GT=0|0
16050679  GT=0|0

 $bcftools view -H chr_22_HG00096.vcf.gz | grep 16050678
# 16050678	.	C	T

$bcftools view -H chr_22_HG00096.vcf.gz | grep 16050679
# 16050679	.	G	A

$samtools faidx hs37d5.fa 22:16050678-16050680
# CGC

$samtools faidx hs37d5.fa 22:16050678-16050680| $bcftools consensus -s ${sample} -H 1 chr_${id}_${sample}.vcf.gz 
# CGC

$samtools faidx hs37d5.fa 22:16050678-16050680| $bcftools consensus -s ${sample} -H 2 chr_${id}_${sample}.vcf.gz 
#CGC

16055171
16055172

 $bcftools view -H chr_22_HG00096.vcf.gz | grep 16055171
# 16055171	.	G	A
 $bcftools view -H chr_22_HG00096.vcf.gz | grep 16055172
# 16055172	.	C	A

$bcftools view -H  chr_22_HG00096.vcf.gz | grep VT=INDEL  | awk '{print $2}'  

51054568
51054569

 $bcftools view -H chr_22_HG00096.vcf.gz | grep 51054568
# 22	51054568	.	TA	T 0|0

 $bcftools view -H chr_22_HG00096.vcf.gz | grep 51054569
 #22	51054569	.	ATATATATATATATATTAGCT	A 0|0

 $samtools faidx hs37d5.fa 22:51054568-51054580
# TATATATATATAT

$samtools faidx hs37d5.fa 22:51054568-51054580| $bcftools consensus -s ${sample} -H 1 chr_${id}_${sample}.vcf.gz 
# TATATATATATAT

$samtools faidx hs37d5.fa 22:51054568-51054580| $bcftools consensus -s ${sample} -H 1 chr_${id}_${sample}.vcf.gz 
# TATATATATATAT

 $bcftools view -H chr22.vcf.gz | grep 51054569




id=22
sample=HG00096
ref=hs37d5

$bcftools view -H chr_22_HG00096.vcf.gz | grep 51054568
# 22	51054568	.	TA	T 0|0

 $bcftools view -H chr_22_HG00096.vcf.gz | grep 51054569
 #22	51054569	.	ATATATATATATATATTAGCT	A 0|0

$samtools faidx hs37d5.fa 22:51054568-51054600
# >22:51054568-51054600
# TATATATATATATATATTAGCTGGGTGTGGTGG

$amtools faidx hs37d5.fa 22:51054568-51054600| $bcftools consensus -s ${sample} -H 1 chr_${id}_${sample}.vcf.gz
# >22:51054568-51054600
# TATATATATATATATATTAGCTGGGTGTGGTGG

$samtools faidx hs37d5.fa 22:51054568-51054600| $bcftools consensus -s ${sample} -H 2 chr_${id}_${sample}.vcf.gz
# >22:51054568-51054600
# TATATATATATATATATTAGCTGGGTGTGGTGG

 $bcftools view -H -i 'POS=51054569 && GT ="1|0"' chr22.vcf.gz 


#************************************************************


# I found the following case:
 $bcftools view -H chr_22_HG00096.vcf.gz | nl | grep "1|0"
# 313579	22	26831676	.	TAAAAGAAAAGAAAAGAAAAG	T	100	PASS	AC=1;AF=0.697484;AN=2;NS=2504;DP=21140;EAS_AF=0.628;AMR_AF=0.7046;AFR_AF=0.8555;EUR_AF=0.7028;SAS_AF=0.545;VT=INDEL	GT	1|0
# 313609	22	26832808	.	ATG	ATGTG,A	100	PASS	AC=1,0;AF=0.0577077,0.336861;AN=2;NS=2504;DP=21130;EAS_AF=0.0169,0.3879;AMR_AF=0.0403,0.3242;AFR_AF=0.0862,0.3865;EUR_AF=0.0845,0.2873;SAS_AF=0.046,0.2771;VT=INDEL;MULTI_ALLELIC	GT	1|0


id=22
sample=HG00096
ref=hs37d5


$samtools faidx hs37d5.fa 22:26831676-26832820
# >22:26831676-26832820
# TAAAAGAAAAGAAAAGAAAAGAAAAGAAAAGAAAAGAAAAGAAAAGAAAAGAAAAATATA
# ATGAGATTTGAGCAATTGACATGGTAGGAATTTCAAGAGTAAAGGGCGGACTGGGTGGAG
# GCTTTGGGGGAGGTGGCCTGAATTCCTGGTCAGCAGTCAGCCGTGAGACTTTGAGCAGGT
# CACTGAGTCTCTCTCTGCCTCAGTTCTTCATCTGTAAAATGGGGAGATGGTACTTACTTC
# ACAGGATTGTTGAGAGGATTCAATGAGTGGCATCTTACATGGCCCCTTGGCAGGATTCCT
# GGCACGTGACAGCATTTGATGAATGGGAGCTAGTAGCTATTATCAAGGACTGTGTTGGAC
# CATGTGAAAAATGATCCATAAGACATAGACCCTGCCCCCAGGAAACAGTGGGAAAGGAAC
# AGGCTTCGGAAGCAAACAGATCTGGACTCAGGTCCTGGCCTCGCCCCATTTGCTGAGTGA
# CCTTGGGCAGGGCTCTGAACCTCCTTAAGCCTCCGTTTTATCTTCTCTAAAACAGAGATG
# ACGATTCCCACTTGCTGGCTGTTTTGAGGAGTGGGAAGTAATATGTGTAAAATGCTTGGC
# AGGTAGAGGGCAGTGGACAGTAAGCTAATCTAAGTGATGGGATGATAATCCAGGTGATGG
# TCATTAGCAAAACTGCTGTGCGAGGCCCCAAGCCTCACTTGGCTGCAGACGGGGTGCCCT
# GGGGAGGTGGGTCTGACACTGGAGGATCCCCACACAGAAATCAGGTGCCGCCAGGGTGGA
# GCGAGGGAGCGGAATCATCCCCTGGGGCTTGGGGAAGCCGGGTCTAGGCTTGCTGGGGAG
# GCCAGTAGTGTGGGCTGAGGTTGGACAGGTCAGGTGGGAGCAAACCTCAGAGGCCCTTGA
# TGCCAGGCTTGAGAATGGAGGCCGTGGGCAATGGGGAGTGATGCTCTTCTTTTGAAAGGT
# TTGCCTTTATGTATTGTTTTAAAACTTCAGAATGACATCTGGGGAAGGGCATTTGAGGAC
# TGATGTGGTAGAATTTCAAAGAGGTGTGAGTAGGAAGTGATGGTAGCAAGGAAAATCTGG
# GGTTTGCTTTCTGTAGATTAACACTTGTTGTTCATCTGTATGATATGCATGCATGTGTGT
# GTGTG


$samtools faidx hs37d5.fa  22:26831676-26832820| $bcftools consensus -s ${sample} -H 1 chr_${id}_${sample}.vcf.gz
# >22:26831676-26832820
# TAAAAGAAAAGAAAAGAAAAGAAAAGAAAAGAAAAATATAATGAGATTTGAGCAATTGAC
# ATGGTAGGAATTTCAAGAGTAAAGGGCGGACTGGGTGGAGGCTTTGGGGGAGGTGGCCTG
# AATTCCTGGTCAGCAGTCAGCCGTGAGACTTTGAGCAGGTCACTGAGTCTCTCTCTGCCT
# CAGTTCTTCATCTGTAAAATGGGGAGATGGTACTTACTTCACAGGATTGTTGAGAGGATT
# CAATGAGTGGCATCTTACATGGCCCCTTGGCAGGATTCCTGGCACGTGACAGCATTTGAT
# GAATGGGAGCTAGTAGCTATTATCAAGGACTGTGTTGGACCATGTGAAAAATGATCCATA
# AGACATAGACCCTGCCCCCAGGAAACAGTGGGAAAGGAACAGGCTTCGGAAGCAAACAGA
# TCTGGACTCAGGTCCTGGCCTCGCCCCATTTGCTGAGTGACCTTGGGCAGGGCTCTGAAC
# CTCCTTAAGCCTCCGTTTTATCTTCTCTAAAACAGAGATGACGATTCCCACTTGCTGGCT
# GTTTTGAGGAGTGGGAAGTAATATGTGTAAAATGCTTGGCAGGTAGAGGGCAGTGGACAG
# TAAGCTAATCTAAGTGATGGGATGATAATCCAGGTGATGGTCATTAGCAAAACTGCTGTG
# CGAGGCCCCAAGCCTCACTTGGCTGCAGACGGGGTGCCCTGGGGAGGTGGGTCTGACACT
# GGAGGATCCCCACACAGAAATCAGGTGCCGCCAGGGTGGAGCGAGGGAGCGGAATCATCC
# CCTGGGGCTTGGGGAAGCCGGGTCTAGGCTTGCTGGGGAGGCCAGTAGTGTGGGCTGAGG
# TTGGACAGGTCAGGTGGGAGCAAACCTCAGAGGCCCTTGATGCCAGGCTTGAGAATGGAG
# GCCGTGGGCAATGGGGAGTGATGCTCTTCTTTTGAAAGGTTTGCCTTTATGTATTGTTTT
# AAAACTTCAGAATGACATCTGGGGAAGGGCATTTGAGGACTGATGTGGTAGAATTTCAAA
# GAGGTGTGAGTAGGAAGTGATGGTAGCAAGGAAAATCTGGGGTTTGCTTTCTGTAGATTA
# ACACTTGTTGTTCATCTGTATGATATGCATGCATGTGTGTGTGTGTG


# Let's try from the haplotype sequence
$samtools faidx  chr_${id}_${sample}_haplotype_1.fa 1:26831676-26832820
>1:26831676-26832820
TTTTGAGACAGAGTCTTGCTCTGTCGCCCAGGCTAGAGTACAGTGGTGCAATCTCGGCTC
ACTGCAACCTCTGCCTCCCGGCTTCAAGCAATTCTCCTGCCTCAGCCTCATGAGTAGTTG
GGATTACAGGCACCCGCCACCATGCCCGGCTAATTTTTCTATTTTTTTTAGTAGAGACGG
GATTTCACCATCTTGGCCAGGCTGGTCTCGAACTCCTGACCTCATGATCCACCTGCCTCA
GCCTCCCAAAGTGCTGGGATTACAGGCGTGAGCCACTGTGCCTGGCCCCAGTAGTGTATT
TGTTTCTAACTTTCTAATTCTCTCATTCTTTTGGCATTCATTAGCTCAGTATAAAGAACT
TTTTGAGAATTTTTTTTTTGAGACAGAGTCTTGCTCTGTTGCCCAGGCTGGAGTGCAGTG
GCACGATCTTGGCTCACTGCAACCTCCGCCTCCAGGTTCAAGCAATTCTCTTGCCTCAGC
CTTCCGAGTAACTGGGATTACAGCCACCTGCCACTACGCCTGAGTAATTTTAGTGGAGAT
GTGGTTTCACCATGTTGGCCAGGCTAGTCTCAAACTCCTGACCTCAAGTGATGTACCTGC
CTTGGCCTCCCAAAGTACTGGGATTACAGGTGTGAGCCACCGCACCCAGACTCAGTAGTG
TATTTTACCTTGTTTTTCTGCCCCTTGTGTTTCCAGCAAACTGAAAGTTAGATCCAGAAG
CTTGATTACATTTAAGAATGCCTGAATGTTTCTTTTTATTTATTTCTTTTTTTTTTTTTT
TTGAAACAAGGTCTCACTCTGTCGCCCAGGCTGGAGTGCAGTGGCACAATCATGGCTCAC
TACAGCCTTGACCTCCTGCCTCAGCCTCCCAAGTAGCTGTGACTACAGGCACGCCATCAC
ACCTGGCTACTTTTTTTTTGAGATGGAGTCTCACTCTGTCGCCCAGGTTGGAGTGCAGTG
GCGTGATCTTGGCTCATTGCAAACTCCGCCTCCCAGGTTCAAGCAATTCTTCTGCCTTAG
CCTCCCGAGTAGCTGGGACTACAGGCACCCGCCACCAAGCCTGGCTAATTTTTGTATTTT
TAGTAGAGATGGGGTTTCACCATATTGGCCAGGCTGGTCTCGAAGGCCTGACTTCGTGAT
CCACC



$samtools faidx hs37d5.fa  22:26831676-26832820| $bcftools consensus -s ${sample} -H 2 chr_${id}_${sample}.vcf.gz

# >22:26831676-26832820
# TAAAAGAAAAGAAAAGAAAAGAAAAGAAAAGAAAAGAAAAGAAAAGAAAAGAAAAATATA
# ATGAGATTTGAGCAATTGACATGGTAGGAATTTCAAGAGTAAAGGGCGGACTGGGTGGAG
# GCTTTGGGGGAGGTGGCCTGAATTCCTGGTCAGCAGTCAGCCGTGAGACTTTGAGCAGGT
# CACTGAGTCTCTCTCTGCCTCAGTTCTTCATCTGTAAAATGGGGAGATGGTACTTACTTC
# ACAGGATTGTTGAGAGGATTCAATGAGTGGCATCTTACATGGCCCCTTGGCAGGATTCCT
# GGCACGTGACAGCATTTGATGAATGGGAGCTAGTAGCTATTATCAAGGACTGTGTTGGAC
# CATGTGAAAAATGATCCATAAGACATAGACCCTGCCCCCAGGAAACAGTGGGAAAGGAAC
# AGGCTTCGGAAGCAAACAGATCTGGACTCAGGTCCTGGCCTCGCCCCATTTGCTGAGTGA
# CCTTGGGCAGGGCTCTGAACCTCCTTAAGCCTCCGTTTTATCTTCTCTAAAACAGAGATG
# ACAATTCCCACTTGCTGGCTGTTTTGAGGAGTGGGAAGTAATATGTGTAAAATGCTTGGC
# AGGTAGAGGGCAGTGGACAGTAAGCTAATCTAAGTGATGGGATGATAATCCAGGTGATGG
# TCATTAGCAAAACTGCTGTGCGAGGCCCCAAGCCTCACTTGGCTGCAGACGGGGTGCCCT
# GGGGAGGTGGGTCTGACACTGGAGGATCCCCACACAGAAATCAGGTGCCGCCAGGGTGGA
# GCGAGGGAGCGGAATCATCCCCTGGGGCTTGGGGAAGCCGGGTCTAGGCTTGCTGGGGAG
# GCCAGTAGTGTGGGCTGAGGTTGGACAGGTCAGGTGGGAGCAAACCTCAGAGGCCCTTGA
# TGCCAGGCTTGAGAATGGAGGCCGTGGGCAATGGGGAGTGATGCTCTTCTTTTGAAAGGT
# TTGCCTTTATGTATTGTTTTAAAACTTCAGAATGACATCTGGGGAAGGGCATTTGAGGAC
# TGATGTGGTAGAATTTCAAAGAGGTGTGAGTAGGAAGTGATGGTAGCAAGGAAAATCTGG
# GGTTTGCTTTCTGTAGATTAACACTTGTTGTTCATCTGTATGATATGCATGCATGTGTGT
# GTGTG


wget http://ftp.ensembl.org/pub/grch37/current/gff3/homo_sapiens/Homo_sapiens.GRCh37.87.chromosome.22.gff3.gz
#https://samtools.github.io/bcftools/howtos/csq-calling.html

$bcftools csq -f hs37d5.fa -g Homo_sapiens.GRCh37.87.chromosome.22.gff3.gz chr22.vcf -Ob -o chr22_hap.bcf

# index the fasta file of choromosome 22 
$samtools faidx linear_bc_chr22.fa
#22:16050075-51244237	35194163	22	60	61

cat linear_bc_chr22.fa | head -10
# >22:16050075-51244237
# AGTGGGCCTAAGTGCCTCCTCTCGGGACTGGTATGGGGACGGTCATGCAATCTGGACAAC
# ATTCACCTTTAAAAGTTTATTGATCTTTTGTGACATGCACGTGGGTTCCCAGTAGCAAGA
# AACTAAAGGGTCGCAGGCCGGTTTCTGCTAATTTCTTTAATTCCAAGACAGTCTCAAATA
# TTTTCTTATTAACTTCCTGGAGGGAGGCTTATCATTCTCTCTTTTGGATGATTCTAAGTA
# CCAGCTAAAATACAGCTATCATTCATTTTCCTTGATTTGGGAGCCTAATTTCTTTAATTT
# AGTATGCAAGAAAACCAATTTGGAAATATCAACTGTTTTGGAAACCTTAGACCTAGGTCA
# TCCTTAGTAAGATCTTCCCATTTATATAAATACTTGCAAGTAGTAGTGCCATAATTACCA
# AACATAAAGCCAACTGAGATGCCCAAAGGGGGCCACTCTCCTTGCTTTTCCTCCTTTTTA
# GAGGATTTATTTCCCATTTTTCTTAAAAAGGAAGAACAAACTGTGCCCTAGGGTTTACTG

# indexing the linear backbonefile
$samtools faidx linear_bc_chr22.fa  22:16050075-51244237:1-20
#AGTGGGCCTAAGTGCCTCCT

$samtools faidx hs37d5.fa 22:16050075-51244237



 $bcftools view -H  $MyVariants | grep VT=INDEL  | awk '{print $2}' | tail -10

 $bcftools view -H  $MyVariants | grep VT=SNP  | awk '{print $2}' | tail -10